census_view.cpp

```diff
@@ -932,6 +936,844 @@ inline std::vector<std::string> const& CensusViewDataViewModel::all_headers() co
+
+    std::string const& col_name(int col) const;
+
+    Input& row_at(int row);
+    Input const& row_at(int row) const;
+
+    any_member<Input>& cell_at(int row, int col);
+    any_member<Input> const& cell_at(int row, int col) const;
+
+  private:
+      any_member<Input>& cell_at(int row, std::string const& col);
+      any_member<Input> const& cell_at(int row, std::string const& col) const;
+      std::vector<std::string> const& all_headers() const;
+
+    CensusGridView& view_;
+};
+
+int CensusViewTable::GetNumberRows()
+{
+    return lmi::ssize(view_.cell_parms());
+}
+
+int CensusViewTable::GetNumberCols()
+{
+    // "+ 1" for cell serial number in first column.
+    return lmi::ssize(view_.visible_columns_) + 1;
+}
+
+wxString CensusViewTable::GetValue(int row, int col)
+{
+    if(col == Col_CellNum)
+        {
+        return value_cast<std::string>(1 + row);
+        }
+
+    auto const& cell = cell_at(row, col);
+    auto const& conv = table_type_converter::get_by_value(cell);
+    return conv.to_renderer_value(cell.str());
+}
+
+void CensusViewTable::SetValue(int row, int col, wxString const& value)
+{
+    LMI_ASSERT(col != Col_CellNum);
+
+    auto& cell = cell_at(row, col);
+    auto const& conv = table_type_converter::get_by_value(cell);
+    auto const& prev_val = cell.str();
+    auto const& new_val = conv.from_editor_value(value);
+
+    if(prev_val == new_val)
+        {
+        return;
+        }
+
+    cell = new_val;
+
+    Input& model = view_.cell_parms()[row];
+    model.Reconcile();
+
+    view_.document().Modify(true);
+}
+
+wxString CensusViewTable::GetTypeName(int row, int col)
+{
+    if(col == Col_CellNum)
+        {
+        return wxGRID_VALUE_NUMBER;
+        }
+
+    auto const& value = cell_at(row, col);
+    auto const& conv = table_type_converter::get_by_value(value);
+
+    auto type = conv.grid_value_type(value);
+
+    return type;
+}
+
+wxString CensusViewTable::GetColLabelValue(int col)
+{
+    if(col == Col_CellNum)
+        {
+        return "Cell";
+        }
+
+    auto const& header = all_headers()[view_.visible_columns_[col - 1]];
+    return insert_spaces_between_words(header);
+}
+
+bool CensusViewTable::AppendRows(size_t numRows)
+{
+    auto grid = GetView();
+    LMI_ASSERT(grid != nullptr);
+
+    wxGridTableMessage msg
+        (this
+        ,wxGRIDTABLE_NOTIFY_ROWS_APPENDED
+        ,bourn_cast<int>(numRows)
+        );
+    grid->ProcessTableMessage(msg);
+
+    return true;
+}
+
+bool CensusViewTable::DeleteRows(size_t pos, size_t num_rows)
+{
+    auto grid = GetView();
+    LMI_ASSERT(grid != nullptr);
+
+    auto const ipos = bourn_cast<int>(pos);
+    auto inum_rows = bourn_cast<int>(num_rows);
+    auto const cur_num_rows = GetView()->GetNumberRows();
+    if(ipos >= cur_num_rows)
+        {
+        wxFAIL_MSG(wxString::Format
+            (wxT("Called CensusViewTable::DeleteRows(pos=%d, N=%d)\nPos value is invalid for present table with %d rows")
```

Comment:
`wxT` shouldn't be used.

But I wonder why isn't `LMI_ASSERT` used here instead of `wxFAIL_MSG` in the first place?